concurrency/ConnectionMultiplexerManager.java

```diff
@@ -0,0 +1,385 @@
+                sessions = sessionsByManager.get(connectionManagerDomain);
+                if (sessions == null) {
+                    sessions = new ConcurrentHashMap<>();
+                    sessionsByManager.put(connectionManagerDomain, sessions);
+                }
+            }
+            /*** END amazon modifications (DCL Security fix ***/
+            sessions.put(streamID, session);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Closes an existing client session that was established through a connection manager.
+     *
+     * @param connectionManagerDomain the connection manager that is handling the connection
+     *        of the session.
+     * @param streamID the stream ID created by the connection manager for the session.
+     */
+    public void closeClientSession(String connectionManagerDomain, String streamID) {
+        Map<String, LocalClientSession> sessions = sessionsByManager.get(connectionManagerDomain);
+        if (sessions != null) {
+            Session session = sessions.remove(streamID);
+            if (session != null) {
+                // Close the session
+                session.close();
+            }
+        }
+    }
+
+    /**
+     * A connection manager has become available. Clients can now connect to the server through
+     * the connection manager.
+     *
+     * @param connectionManagerName the connection manager that has become available.
+     */
+    public void multiplexerAvailable(String connectionManagerName) {
+        // Add a new entry in the list of available managers. Here is where we are going to store
+        // which clients were connected through which connection manager
+        /*** BEGIN amazon modifications (DCL Security fix ***/
+        synchronized (connectionManagerName.intern()) {
+            Map<String, LocalClientSession> sessions = sessionsByManager.get(connectionManagerName);
+            if (sessions == null) {
+                sessions = new ConcurrentHashMap<>();
+                sessionsByManager.put(connectionManagerName, sessions);
+            }
+        }
+        /*** END amazon modifications (DCL Security fix ***/
+    }
+
+    /**
+     * A connection manager has gone unavailable. Close client sessions that were established
+     * to the specified connection manager.
+     *
+     * @param connectionManagerName the connection manager that is no longer available.
+     */
+    public void multiplexerUnavailable(String connectionManagerName) {
+        // Remove the connection manager and the hosted sessions
+        Map<String, LocalClientSession> sessions = sessionsByManager.remove(connectionManagerName);
+        if (sessions != null) {
+            for (String streamID : sessions.keySet()) {
+                // Remove inverse track of connection manager hosting streamIDs
+                streamIDs.remove(streamID);
+                // Close the session
+                sessions.get(streamID).close();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.